dvc/external_repo.py

```diff
@@ -139,110 +140,143 @@ def download_update(result):
-            # same as if it was a remote DVC repo.
             return
 
         try:
-            remote_name = self.config["core"].get("remote")
+            remote_name = repo.config["core"].get("remote")
             if remote_name:
-                self._fix_local_remote(src_repo, remote_name)
+                self._fix_local_remote(repo, src_repo, remote_name)
             else:
-                self._add_upstream(src_repo)
+                self._add_upstream(repo, src_repo)
         finally:
             src_repo.close()
 
-    def _fix_local_remote(self, src_repo, remote_name):
-        # If a remote URL is relative to the source repo,
-        # it will have changed upon config load and made
-        # relative to this new repo. Restore the old one here.
-        new_remote = self.config["remote"][remote_name]
-        old_remote = src_repo.config["remote"][remote_name]
-        if new_remote["url"] != old_remote["url"]:
-            new_remote["url"] = old_remote["url"]
+    def _get_cache_dir(self):
+        try:
+            cache_dir = CACHE_DIRS[self.url]
+        except KeyError:
+            cache_dir = CACHE_DIRS[self.url] = tempfile.mkdtemp("dvc-cache")
+        return cache_dir
 
-    def _add_upstream(self, src_repo):
-        # Fill the empty upstream entry with a new remote pointing to the
-        # original repo's cache location.
-        cache_dir = src_repo.cache.local.cache_dir
-        self.config["remote"]["auto-generated-upstream"] = {"url": cache_dir}
-        self.config["core"]["remote"] = "auto-generated-upstream"
+
+class ExternalRepo(Repo, BaseExternalRepo):
+    def __init__(
+        self,
+        root_dir,
+        url,
+        scm=None,
+        rev=None,
+        for_write=False,
+        cache_dir=None,
+        cache_types=None,
+        **kwargs
+    ):
+        super().__init__(root_dir, scm=scm, rev=rev)
+
+        self.url = url
+        self.for_write = for_write
+        self.cache_dir = cache_dir or self._get_cache_dir()
+        self.cache_types = cache_types
+
+        self._setup_cache(self)
+        self._fix_upstream(self)
+        self.tree_confs = kwargs
 
 
 class ExternalGitRepo(BaseExternalRepo):
-    def __init__(self, root_dir, url, rev):
+    def __init__(
+        self,
+        root_dir,
+        url,
+        scm=None,
+        rev=None,
+        for_write=False,
+        cache_dir=None,
+        cache_types=None,
+        **kwargs
+    ):
         self.root_dir = os.path.realpath(root_dir)
+        self.scm = scm
+
         self.url = url
+        self.for_write = for_write
+        self.cache_dir = cache_dir or self._get_cache_dir()
+        self.cache_types = cache_types
         self.tree = self.scm.get_tree(rev)
+        self.tree_confs = kwargs
 
-    @cached_property
-    def scm(self):
-        return Git(self.root_dir)
+        self.config = {"cache": {"dir": self.cache_dir}}
+        self.cache = Cache(self)
+        self.state = StateNoop()
```

Comment:
The biggest issues/controversial thing could be these two. I tried to unify them though, rather than trying to avoid, as most of the trees already try to access `repo.cache`, which the cache then tries to use `State`.

Reply:
Hmm, maybe we should have something that will be a common ancestor between  `Repo` and `BaseExternalRepo`? 
Modeling `ExternalGitRepo` to partially match the structure of `Repo` is probably not good for the long run.

Reply:
@pared, let's duck type, for now. Let's reconsider in the future.